daffodil-core/src/main/scala/org/apache/daffodil/grammar/ElementBaseGrammarMixin.scala

```diff
@@ -17,132 +17,115 @@
+import org.apache.daffodil.grammar.primitives.LeftCenteredPadding
+import org.apache.daffodil.grammar.primitives.LiteralCharacterNilOfSpecifiedLength
+import org.apache.daffodil.grammar.primitives.LiteralNilDelimitedEndOfData
 import org.apache.daffodil.grammar.primitives.LiteralValueNilOfSpecifiedLength
-import org.apache.daffodil.schema.annotation.props.NotFound
+import org.apache.daffodil.grammar.primitives.LogicalNilValue
+import org.apache.daffodil.grammar.primitives.NilLiteralCharacter
+import org.apache.daffodil.grammar.primitives.OVCRetry
+import org.apache.daffodil.grammar.primitives.OnlyPadding
+import org.apache.daffodil.grammar.primitives.PackedDecimalDelimitedEndOfData
+import org.apache.daffodil.grammar.primitives.PackedDecimalKnownLength
+import org.apache.daffodil.grammar.primitives.PackedDecimalRuntimeLength
+import org.apache.daffodil.grammar.primitives.PackedIntegerDelimitedEndOfData
+import org.apache.daffodil.grammar.primitives.PackedIntegerKnownLength
+import org.apache.daffodil.grammar.primitives.PackedIntegerRuntimeLength
+import org.apache.daffodil.grammar.primitives.PaddingInfoMixin
+import org.apache.daffodil.grammar.primitives.PrefixLength
+import org.apache.daffodil.grammar.primitives.RightCenteredPadding
+import org.apache.daffodil.grammar.primitives.RightFill
+import org.apache.daffodil.grammar.primitives.SimpleNilOrValue
+import org.apache.daffodil.grammar.primitives.SpecifiedLengthExplicit
+import org.apache.daffodil.grammar.primitives.SpecifiedLengthExplicitCharacters
+import org.apache.daffodil.grammar.primitives.SpecifiedLengthImplicit
+import org.apache.daffodil.grammar.primitives.SpecifiedLengthImplicitCharacters
+import org.apache.daffodil.grammar.primitives.SpecifiedLengthPattern
+import org.apache.daffodil.grammar.primitives.StringDelimitedEndOfData
+import org.apache.daffodil.grammar.primitives.StringOfSpecifiedLength
+import org.apache.daffodil.grammar.primitives.Terminator
+import org.apache.daffodil.grammar.primitives.TrailingSkipRegion
+import org.apache.daffodil.grammar.primitives.ZonedTextIntPrim
+import org.apache.daffodil.processors.TextJustificationType
 import org.apache.daffodil.schema.annotation.props.Found
+import org.apache.daffodil.schema.annotation.props.NotFound
+import org.apache.daffodil.schema.annotation.props.gen.BinaryFloatRep
+import org.apache.daffodil.schema.annotation.props.gen.BinaryNumberRep
+import org.apache.daffodil.schema.annotation.props.gen.LengthKind
+import org.apache.daffodil.schema.annotation.props.gen.LengthUnits
+import org.apache.daffodil.schema.annotation.props.gen.NilKind
+import org.apache.daffodil.schema.annotation.props.gen.Representation
+import org.apache.daffodil.schema.annotation.props.gen.TextNumberRep
+import org.apache.daffodil.schema.annotation.props.gen.YesNo
+import org.apache.daffodil.util.PackedSignCodes
+import org.apache.daffodil.xml.GlobalQName
+import org.apache.daffodil.xml.XMLUtils
```

Comment:
What happened here? Where these just autosorted by eclipse? Perhaps we should make it part of our code review process that when we add new imports we add them sorted so that these autoformat changes don't overwhelm patch review?

Reply:
I looked into this Eclipse IDE behavior. Besides organizing the imports that are automatically inserted into groups it also alphabetizes, and there appears to be no way to turn this off for java or scala. You can specify the grouping order for Java and Scala separately. E.g., the order is java, scala, org, com. But there is no setting to get Eclipse to only add them at the bottom and don't re-order. There just isn't a setting for this. You can turn off the whole mechanism,... but this ability to just add all the imports that are missing is very time saving. So I'd prefer not to turn this off. 

So I guess yeah, we should specify that the imports should be alphabetized within groups, and the group order is java, scala, org, com (for scala code), and java, javax, org, com (for java code).

Reply:
Sounds reasonable to me. How does it handle other things. For example, we have a handful of things that start with "net", "unit", "passera", and then some things like "import Maybe._","import NodeInfo._", etc.

Also, out of curiousity, does it do anything different with things in the same package namespace? I.e. since the packages are org.apache.daffodil, does it always put org.apache.daffodil.whatever imports at the top/bottom, or are they sorted just like any other import org?

Reply:
File FNFunctions.scala is a good example:

Before:

import org.apache.daffodil.processors._
import org.apache.daffodil.infoset._
import org.apache.daffodil.processors.unparsers.UnparseError
import org.apache.daffodil.api.Diagnostic
import org.apache.daffodil.api.DataLocation
import org.apache.daffodil.util.Maybe
import Maybe._
import org.apache.daffodil.exceptions._
import com.ibm.icu.util.Calendar
import scala.math.BigDecimal.RoundingMode
import com.ibm.icu.util.TimeZone
import org.apache.daffodil.util.Numbers._
import org.apache.daffodil.calendar.DFDLDateTime
import org.apache.daffodil.calendar.DFDLCalendar
import org.apache.daffodil.calendar.DFDLTime
import org.apache.daffodil.calendar.DFDLDate
import java.lang.{ Number => JNumber, Byte => JByte, Short => JShort, Integer => JInt, Long => JLong, Float => JFloat, Double => JDouble, Boolean => JBoolean }
import java.math.{ BigDecimal => JBigDecimal, BigInteger => JBigInt }

After:

import java.lang.{ Number => JNumber, Byte => JByte, Short => JShort, Integer => JInt, Long => JLong, Float => JFloat, Double => JDouble, Boolean => JBoolean }
import java.math.{ BigDecimal => JBigDecimal, BigInteger => JBigInt }

import scala.math.BigDecimal.RoundingMode

import org.apache.daffodil.api.DataLocation
import org.apache.daffodil.api.Diagnostic
import org.apache.daffodil.calendar.DFDLCalendar
import org.apache.daffodil.calendar.DFDLDate
import org.apache.daffodil.calendar.DFDLDateTime
import org.apache.daffodil.calendar.DFDLTime
import org.apache.daffodil.exceptions._
import org.apache.daffodil.infoset._

import org.apache.daffodil.processors._
import org.apache.daffodil.processors.unparsers.UnparseError
import org.apache.daffodil.util.Maybe
import org.apache.daffodil.util.Maybe._
import org.apache.daffodil.util.Numbers._

import com.ibm.icu.util.Calendar
import com.ibm.icu.util.TimeZone

I do not understand the blank line in the middle of the org.apache imports after infoset._. 

Note that it also undid the import Maybe._, adding the global package name.